### PR TITLE
refactor decorator cache and move to class metadata

### PR DESCRIPTION
**Pull request by captaincaius**
This PR makes progress on #44 ... one singleton down, several more to go ;)

Some of the names don't make perfect sense anymore.  I left them as-is to make your code-review easier.  If you check the diff with whitespace ignored you can see that the original (nice) solution of deferring decorator processing and using recursion to maintain order is preserved exactly.

This PR should also help with a feature request I haven't submitted yet but I've already got some PoC code on ;) (composing and picking from models in a typed fashion).

Let me know if you'd like anything changed.

Thanks! :)

- original decorator untouched but data and code separated
- location of data moved from singleton to class metadata itself
- some residue removed

### CODE REVIEW

File: src/prop.ts [hunk 1]
```diff
@@ -52,241 +45,255 @@ function baseProp(
...
-      virtuals.set(name, new Map());
-    }
-
-    if (utils.isWithVirtualPOP(rawOptions)) {
-      if (!utils.includesAllVirtualPOP(rawOptions)) {
-        throw new NotAllVPOPElementsError(name, key);
-      }
-      virtuals.get(name).set(key, rawOptions);
+  const existingMapForTarget = Reflect.getOwnMetadata(DecoratorKeys.PropProps, target) as DecoratedPropertyMetadataMap;
+  if (!existingMapForTarget) {
```
Comment (hasezoey) [inline src/prop.ts]: i would prefer if you would use `isNullOrUndefined`

**Comment by hasezoey** [comment]
And i will probably *not* merge this into 6.0.0 maybe in a 6.x, because i target to release 6.0.0 on monday (30.9.2019)

**Comment by hasezoey** [comment]
@captaincaius i would want to include this in 6.1.0, after reviewing it again & closer, could you update to latest changes?

File: src/prop.ts (hunk 1 above)
Comment (captaincaius) [reply src/prop.ts]: good catch.  will do :)

**Comment by captaincaius** [comment]
Hi.  Sure, I'd be happy to!

But are you finished reviewing it?  As I mentioned, there's some cleanup/refactoring I'd like to do first - but I didn't do it yet in the interest of making review easier.

Or if you want I can do that in a separate PR - up to you.

Thanks for doing such an outstanding job maintaining this project btw!

**Comment by hasezoey** [comment]
> Or if you want I can do that in a separate PR - up to you.

if they are the same changes as here, then dont create a new one, if the changes are different, feel free to open a new pr (i just dont really like to always have a new pr "when making a change", like in the old repo)

> But are you finished reviewing it?

yes i roughly reviewed, as you could see with my review, otherwise i would review it in detail again when it is finished, try to keep changes that are not related seperate :)

**Comment by captaincaius** [comment]
Ok then I guess I'll just update this, and daisy-chain a PR for the cleanups.

Thanks!

**Comment by captaincaius** [comment]
Synced and resolved merge conflicts again.  Do you plan on merging this?  Even though it's a pretty simple PR diff when ignoring whitespace, because of the indentation changes, resolving merge conflicts is manual and super tedious cause git's misjudging what's changed :(.  Every time someone sneezes on prop.ts git gets confused and makes a mess :D

**Comment by hasezoey** [comment]
> Every time someone sneezes on prop.ts git gets confused and makes a mess :D

i absolutely know what you mean, the DEO change & branch were always messed up

---

>  Do you plan on merging this?

yes i want to, but you said this pr was not final yet?

> Ok then I guess I'll just update this, and daisy-chain a PR for the cleanups.

-> because of this i waited for you to write something like "this pr is now finish / ready to merge"

**Comment by captaincaius** [comment]
Ha.  I just read back and you're totally right.  Sorry I wasn't clear.

It's ready.

There's some naming tasks that can be done, but I'll defer them to followup issues (that I'll probably take) after it's merged.

**Comment by hasezoey** [comment]
I prepared version 6.1.0 (pre) and merged it into here, then fixed the initial style warnings

otherwise i added some tsdoc, renamed some DecoratorKeys and moved the types that you moved to "data.ts" to "types.ts"

and i changed how "baseProp" & "_buildPropMetadata" get their arguments, instead of always having to redefine them

do you think these changes are ok and still ready to merge?

**Comment by captaincaius** [comment]
@hasezoey looks like you beat me on all the cleanups and more! :+1: 

Yes I think it's even more ready to merge now :D.

I think there won't be a need for that cleanup issue anymore.

**Comment by hasezoey** [comment]
@captaincaius you are now free to delete your old/merged branches :)

**Comment by captaincaius** [comment]
woohoo.  hopefully soon I can work on moving the rest of whatever can move to metadata too, unless someone beats me to it.

**Pull request merged**